### PR TITLE
Fix bug with list_artifacts in gcp

### LINKED ISSUE #3240: Error in loading spark ml models from gcs_artifact_storage [BUG]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [x] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **MLflow version (run ``mlflow --version``)**: 1.9.1
- **Python version**: 3.6.10

### Describe the problem
While loading/fetching sparkml models/artifacts from GCS bucket, raises the below error because `client.list_artifacts(run_id, artifact_relative_path)` returns the **<FileInfo: file_size=0, is_dir=False, path=artifact_relative_path +'/'>** which(path) does not exist along with other FileInfos. Hence it sends 404 error for file not found.

### Code to reproduce the issue
`mlflow.spark.load_model(gcs_path_to_model)`
or 
`client.download_artifacts(run_id, sparkml_model_relative_gcs_path)`

### Other info / logs
Complete error logs
2020/07/07 13:49:25 INFO mlflow.spark: URI 'gs://gcs-datascience-adhoc-us/ashwar/ecf191ff5c864bc99b75e531b31c3e4a/artifacts/spark-model/sparkml' does not point to the current DFS.
2020/07/07 13:49:25 INFO mlflow.spark: File 'gs://gcs-datascience-adhoc-us/ashwar/ecf191ff5c864bc99b75e531b31c3e4a/artifacts/spark-model/sparkml' not found on DFS. Will attempt to upload the file.
Traceback (most recent call last):
  File "/opt/conda/default/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 951, in download_to_file
    transport, file_obj, download_url, headers, start, end, raw_download
  File "/opt/conda/default/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 834, in _do_download
    download.consume(transport)
  File "/opt/conda/default/lib/python3.6/site-packages/google/resumable_media/requests/download.py", line 153, in consume
    self._process_response(result)
  File "/opt/conda/default/lib/python3.6/site-packages/google/resumable_media/_download.py", line 171, in _process_response
    response, _ACCEPTABLE_STATUS_CODES, self._get_status_code
  File "/opt/conda/default/lib/python3.6/site-packages/google/resumable_media/_helpers.py", line 96, in require_status_code
    *status_codes
google.resumable_media.common.InvalidResponse: ('Request failed with status code', 404, 'Expected one of', <HTTPStatus.OK: 200>, <HTTPStatus.PARTIAL_CONTENT: 206>)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/ashwar_gupta_zeotap_com/pyspark/predict.py", line 29, in <module>
    model = mlflow.spark.load_model(model_production_uri, "hdfs:/user/mlflow/temp")
  File "/opt/conda/default/lib/python3.6/site-packages/mlflow/spark.py", line 501, in load_model
    return _load_model(model_uri=model_uri, dfs_tmpdir=dfs_tmpdir)
  File "/opt/conda/default/lib/python3.6/site-packages/mlflow/spark.py", line 452, in _load_model
    model_path = _HadoopFileSystem.maybe_copy_from_uri(model_uri, tmp_path)
  File "/opt/conda/default/lib/python3.6/site-packages/mlflow/spark.py", line 302, in maybe_copy_from_uri
    return cls.maybe_copy_from_local_file(_download_artifact_from_uri(src_uri), dst_path)
  File "/opt/conda/default/lib/python3.6/site-packages/mlflow/tracking/artifact_utils.py", line 74, in _download_artifact_from_uri
    artifact_path=artifact_path, dst_path=output_path)
  File "/opt/conda/default/lib/python3.6/site-packages/mlflow/store/artifact/artifact_repo.py", line 128, in download_artifacts
    return download_artifact_dir(artifact_path)
  File "/opt/conda/default/lib/python3.6/site-packages/mlflow/store/artifact/artifact_repo.py", line 106, in download_artifact_dir
    download_file(file_info.path)
  File "/opt/conda/default/lib/python3.6/site-packages/mlflow/store/artifact/artifact_repo.py", line 90, in download_file
    self._download_file(remote_file_path=fullpath, local_path=local_file_path)
  File "/opt/conda/default/lib/python3.6/site-packages/mlflow/store/artifact/gcs_artifact_repo.py", line 105, in _download_file
    gcs_bucket.blob(remote_full_path).download_to_filename(local_path)
  File "/opt/conda/default/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 1024, in download_to_filename
    if_metageneration_not_match=if_metageneration_not_match,
  File "/opt/conda/default/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 954, in download_to_file
    _raise_from_invalid_response(exc)
  File "/opt/conda/default/lib/python3.6/site-packages/google/cloud/storage/blob.py", line 3091, in _raise_from_invalid_response
    raise exceptions.from_http_status(response.status_code, message, response=response)
google.api_core.exceptions.NotFound: 404 GET https://storage.googleapis.com/download/storage/v1/b/gcs-datascience-adhoc-us/o/ashwar%2Fecf191ff5c864bc99b75e531b31c3e4a%2Fartifacts%2Fspark-model%2Fsparkml?alt=media: ('Request failed with status code', 404, 'Expected one of', <HTTPStatus.OK: 200>, <HTTPStatus.PARTIAL_CONTENT: 206>)
20/07/07 13:49:25 INFO org.spark_project.jetty.server.AbstractConnector: Stopped Spark@5b7269db{HTTP/1.1,[http/1.1]}{0.0.0.0:4040}

**PS: I have found the fix and tested the same.**   

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### PR DESCRIPTION
**Pull request by santosh1994**
## What changes are proposed in this pull request?

Fixes bug where `GCSArtifactRepository.list_artifacts` returns incorrect objects list with current directory included as regular file.

Prefix will always be directory since we appending / , And in GCP list_blobs method the current directory is also returned. So we can skip adding current directory to list of regular files
Fixes #3240 

## How is this patch tested?

Test manually for `mlflow.spark.load_model` and `mlflow.sklearn.load_model`

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
It seems that when we create a folder on the cloud storage browser, google cloud storage treats it as an object rather than a folder because google cloud storage does not have folders or subdirectories.

**Comment by harupy** [comment]
@santosh1994 Can you sync with the master branch and resolve conflicts?

**Comment by santosh1994** [comment]
resolved conflicts

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 1]
```diff
@@ -86,6 +86,8 @@ def list_artifacts(self, path=None):
 
         results = bkt.list_blobs(prefix=prefix, delimiter="/")
         for result in results:
+            if result.name == prefix:
```
Comment (harupy) [inline mlflow/store/artifact/gcs_artifact_repo.py:89]: @santosh1994 Can you insert a comment here to explain why we need this? May you could use https://github.com/googleapis/google-cloud-python/issues/5163 as a reference.

Comment (santosh1994) [reply mlflow/store/artifact/gcs_artifact_repo.py:91]: done

**Review by harupy — APPROVED** [review]
LGTM

**Pull request merged**